compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/optimizer/BIRBasicBlockOptimizer.java

```diff
@@ -0,0 +1,311 @@
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package org.wso2.ballerinalang.compiler.bir.optimizer;
+
+import org.wso2.ballerinalang.compiler.bir.model.BIRNode;
+import org.wso2.ballerinalang.compiler.bir.model.BIRNode.BIRBasicBlock;
+import org.wso2.ballerinalang.compiler.bir.model.BIRTerminator;
+import org.wso2.ballerinalang.compiler.bir.model.BIRVisitor;
+import org.wso2.ballerinalang.compiler.util.Name;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Remove unnecessary goto basic blocks.
+ *
+ * @since 2.0
+ */
+
+public class BIRBasicBlockOptimizer extends BIRVisitor {
+
+    private BIROptimizer.OptimizerEnv env;
+    private final Set<BIRBasicBlock> predecessors = new HashSet<>();
+    private final Map<BIRBasicBlock, List<BIRBasicBlock>> predecessorMap = new HashMap<>();
+
+    private int currentBBId = -1;
+    private static final String BIR_BASIC_BLOCK_PREFIX = "bb";
+
+    public void optimizeNode(BIRNode node, BIROptimizer.OptimizerEnv env) {
+        if (node == null) {
+            return;
+        }
+        BIROptimizer.OptimizerEnv oldEnv = this.env;
+        this.env = env;
+        node.accept(this);
+        this.env = oldEnv;
+    }
+
+    public void optimizeTerm(BIRTerminator term, BIROptimizer.OptimizerEnv env) {
+        this.optimizeNode(term, env);
+    }
+
+    @Override
+    public void visit(BIRNode.BIRPackage birPackage) {
+        birPackage.typeDefs.forEach(tDef -> tDef.accept(this));
+        birPackage.functions.forEach(func -> func.accept(this));
+    }
+
+    @Override
+    public void visit(BIRNode.BIRTypeDefinition birTypeDefinition) {
+        birTypeDefinition.attachedFuncs.forEach(func -> func.accept(this));
+    }
+
+    @Override
+    public void visit(BIRNode.BIRFunction birFunction) {
+        BIROptimizer.OptimizerEnv funcEnv = new BIROptimizer.OptimizerEnv();
+
+        // get basic blocks vs predecessors map
+        getPredecessorMap(birFunction.basicBlocks);
+
+        // get trap end basic blocks into a list
+        List<BIRBasicBlock> trapEndBasicBlocks = new ArrayList<>();
+        birFunction.errorTable.forEach(errorEntry -> trapEndBasicBlocks.add(errorEntry.endBB));
+
+        // Remove unreachable basic blocks
+        Set<BIRBasicBlock> unreachableBasicBlocks = new HashSet<>();
```

Comment:
Shall we have multiple small methods for each optimizations and call them in the `visit` method? This seems to be a large method.